test/broadcasting.jl

```diff
@@ -0,0 +1,221 @@
+module TestBroadcasting
+
+using Test, DataFrames
+
+refdf = DataFrame(ones(3, 5))
+
+@testset "normal data frame and data frame row in broadcasted assignment - one column" begin
+    df = copy(refdf)
+    df[1] .+= 1
+    @test df.x1 == [2, 2, 2]
+    @test all(Matrix(df[2:end]) .== ones(size(df[2:end])...))
+
+    dfv = @view df[1:2, 2:end]
+    dfv[1] .+= 1
+    @test dfv.x2 == [2, 2]
+    @test all(Matrix(dfv[2:end]) .== ones(size(dfv[2:end])...))
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    dfr = df[1, 3:end]
+    dfr[end-1:end] .= 10
+    @test all(Vector(dfr) .== [1, 10, 10])
+    @test df.x3[1] == 1 && df.x4[1] == 10 && df.x5[1] == 10
+
+    df = copy(refdf)
+    df[1] .+= [1, 1, 1]
+    @test df.x1 == [2, 2, 2]
+    @test all(Matrix(df[2:end]) .== ones(size(df[2:end])...))
+
+    dfv = @view df[1:2, 2:end]
+    dfv[1] .+= [1, 1]
+    @test dfv.x2 == [2, 2]
+    @test all(Matrix(dfv[2:end]) .== ones(size(dfv[2:end])...))
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    dfr = df[1, 3:end]
+    dfr[end-1:end] .= [10, 10]
+    @test all(Vector(dfr) .== [1, 10, 10])
+    @test df.x3[1] == 1 && df.x4[1] == 10 && df.x5[1] == 10
+
+    df = copy(refdf)
+    dfv = @view df[1:2, 2:end]
+    dfr = df[1, 3:end]
+    @test_throws DimensionMismatch df[1] .= rand(3, 1)
+    @test_throws DimensionMismatch dfv[1] .= rand(2, 1)
+    @test_throws DimensionMismatch dfr[end-1:end] .= rand(3, 1)
+
+    df = copy(refdf)
+    df[:x1] .+= 1
+    @test df.x1 == [2, 2, 2]
+    @test all(Matrix(df[2:end]) .== ones(size(df[2:end])...))
+
+    dfv = @view df[1:2, 2:end]
+    dfv[:x2] .+= 1
+    @test dfv.x2 == [2, 2]
+    @test all(Matrix(dfv[2:end]) .== ones(size(dfv[2:end])...))
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    dfr = df[1, 3:end]
+    dfr[[:x4, :x5]] .= 10
+    @test all(Vector(dfr) .== [1, 10, 10])
+    @test df.x3[1] == 1 && df.x4[1] == 10 && df.x5[1] == 10
+
+    df = copy(refdf)
+    df[:x1] .+= [1, 1, 1]
+    @test df.x1 == [2, 2, 2]
+    @test all(Matrix(df[2:end]) .== ones(size(df[2:end])...))
+
+    dfv = @view df[1:2, 2:end]
+    dfv[:x2] .+= [1, 1]
+    @test dfv.x2 == [2, 2]
+    @test all(Matrix(dfv[2:end]) .== ones(size(dfv[2:end])...))
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    dfr = df[1, 3:end]
+    dfr[[:x4, :x5]] .= [10, 10]
```

Comment:
This is one of the crucial points of the design decision. If `dfr` is to be one-dimensional it has to be treated in assignment as column vector as this is what we have in Base. This is related to the fact that `axes(dfr)` returns one element tuple, not two-element tuple with first element being a `Base.OneTo(1)`.

If we want to change this we have to make `DataFrameRow` have `ndims` equal to 2. But then the question is how would it be different from a `DataFrame` with one row (it would not differ except 
 for indexing rules). If we went this way then actually I think we would have to drop having `DataFrameRow` altogether.

Put it otherwise. This is the same as with arrays in Base:

```
julia> x = ones(2,3)
2×3 Array{Float64,2}:
 1.0  1.0  1.0
 1.0  1.0  1.0

julia> x[1,:]
3-element Array{Float64,1}:
 1.0
 1.0
 1.0

julia> x[1, :] .= [10, 10, 10]
3-element view(::Array{Float64,2}, 1, :) with eltype Float64:
 10.0
 10.0
 10.0

julia> x[1, :] .= [10 10 10]
ERROR: DimensionMismatch("cannot broadcast array to have fewer dimensions")
```

Reply:
I leave it open for a discussion.

Reply:
@nalimilan - how do you see it (I would like to make a decision and then decide on https://github.com/JuliaData/DataFrames.jl/issues/1806 after we merge this PR). Then I will move to `setindex!`. The ultimate objective is to have a release relatively soon so that I can present a "cleaned up" and released functionality during JuliaCon2019, as I run a tutorial on DataFrames.jl there.

Reply:
Interesting. Both approaches have some merit. Making `DataFrameRow` act like a single-row matrix kind of makes sense, but  that wouldn't be consistent with array indexing in Base: `A[1, :]` returns a vector. That would also be inconsistent with `NamedTuple`, and as I noted at https://github.com/JuliaData/DataFrames.jl/issues/1806 I think we want to allow broadcasting these two types together.

Anyway I think I misread the line and interpreted it as `df[[:x4, :x5]] .= [10, 10]` recycling `[10, 10]` over rows, which would be much more problematic.

In the `dfr` case, what's the advantage of `.=` over just `=`? Maybe we could just disallow broadcasting vectors and matrices with `DataFrameRow` for now? Only scalar broadcasting sounds useful.

Reply:
OK - I will restrict broadcasting of `DataFrameRow` to scalars for now.
Actually this is related to #1806. I will comment on it there (but what you propose is safe to do).

Reply:
I have thought about it more and my conclusion is that I would rather allow vectors.

First regarding

> Only scalar broadcasting sounds useful.

```
dfr .= 1 .+ [1,2,3] .+ sin.([5,6,7])
```

will not allocate, while

```
dfr = 1 .+ [1,2,3] .+ sin.([5,6,7])
```

will allocate.

Secondly - if we think of `DataFrameRow` as a kind of mutable `NamedTuple` it is hard to imagine it has different dimensionality than `1`.

Thirdly (this is a minor issue) - in order to restrict ourselves to scalars `DataFrameRow` would have to pretend being a scalar (which would in some corner cases produce strange errors).

Let us think about it a bit more. I will shortly open another PR regarding to `DataFrameRow` design that I think should be discussed before we move forward.

Reply:
Also we define `Vector(dfr)`

Reply:
Fine with me. It's a bit weird that a data frame row behaves like a column vector, but it's not our fault if vectors are considered as columns in Julia.

Reply:
OK - so I will leave it as is and emphasize in the documentation. Actually I want to make `DataFrameRow` behave like a mutable `NamedTuple` (but it is also considered 1-dimensional, of course it is not mutable). See https://github.com/JuliaData/DataFrames.jl/pull/1813 for additional changes that improve this consistency.